Incoming tag sets should be replaced on a namespace (RHCLOUD-5156)

Glutexo opened the pull request:
Allows passing tags in host ingress in the nested format. That allows to delete a namespace by including it in the object, not listing any actual tags for it. The original structured format is still supported. REST API remains unchanged, still supporting only the structured format.

Resolves [_RHCLOUD-5156_](https://projects.engineering.redhat.com/browse/RHCLOUD-5156).

## Steps to reproduce:

1. Create a host with some tags. `{"namespace": {"key": ["value"]}}`
2. Update the host with a set of tags that contains the namespace, but no values. `{"namespace": {"key": []}}`
3. Get the host and see it has no tags under that namespace.

Glutexo commented:
One failing test in every file. A winning sequence.

jharting commented on app/models.py:
```diff
@@ -336,6 +347,47 @@ def has_timezone_info(self, timestamp):
+                    raise ValidationError("Tag values must be an array or null.")
+
+                for value in values:
+                    if value is None:
```
I think we should validate that "values" is a list of **non-null** values. See my comment at https://projects.engineering.redhat.com/browse/RHCLOUD-5805

Glutexo replied:
It’s been like this ever since. Both OpenAPI [specification](https://github.com/RedHatInsights/insights-host-inventory/blob/23eb955783381e544e5bef50230be01b82ccc29e/swagger/api.spec.yaml#L518) and _TagsSchema_ explicitly [allowed](https://github.com/RedHatInsights/insights-host-inventory/blob/23eb955783381e544e5bef50230be01b82ccc29e/app/models.py#L312) _null_ value. These are [then](https://github.com/RedHatInsights/insights-host-inventory/blob/23eb955783381e544e5bef50230be01b82ccc29e/app/utils.py#L301) [ignored](https://github.com/RedHatInsights/insights-host-inventory/blob/23eb955783381e544e5bef50230be01b82ccc29e/app/utils.py#L310) and the key is saved without the empty value.

It had to be like this. Otherwise it wouldn’t be possible to add a value-less tag using the structured format: `{"namespace": "namespace", "key": "key", "value": null}`

Glutexo commented on app/models.py:
```diff
@@ -98,12 +102,13 @@ def __init__(
         self.reporter = reporter
 
     def save(self):
+        self._cleanup_tags()
```
This may look fishy as this effectively introduces state. 😱 That’s however unfortunately an inseparable feature of:

- Using a DB model class for business logic; during a save, this already happens.
- Using the same object for the input data (not exactly a host representation, more a create host request) and for the created/updated host and its output representation.

During the [update](https://github.com/Glutexo/insights-host-inventory/blob/32ad5800d604400a70e3efb19ddae74893c630f5/lib/host_repository.py#L156), this is better, because we use the input host only as a set of new host data / update commands. Here, on [create](https://github.com/Glutexo/insights-host-inventory/blob/32ad5800d604400a70e3efb19ddae74893c630f5/lib/host_repository.py#L145) this is worse, because after the save, the input host object becomes a created host object. As already stated, it’s just how things are.

Glutexo commented on app/models.py:
```diff
@@ -189,19 +194,25 @@ def replace_facts_in_namespace(self, namespace, facts_dict):
 
-            for input_namespace, input_tags in tags_dict.items():
-                self.replace_tags_in_namespace(input_namespace, input_tags)
+        for namespace in namespaces_to_delete:
```
Not doing this the original iteration, because it would break if _self.tags_ was passed as _ tags_dict_. Not really happening, just a pre-caution.

jharting replied:
It's not clear to my why this would break. Can you elaborate and perhaps add it as a code comment also so that the second loop does not get removed on next refactoring?

Glutexo commented on app/models.py:
```diff
@@ -189,19 +194,25 @@ def replace_facts_in_namespace(self, namespace, facts_dict):
-        self.tags[namespace] = tags_dict
         orm.attributes.flag_modified(self, "tags")
 
+    def _cleanup_tags(self):
+        self._update_tags(self.tags)
+
```
For the sake of semantics instead of explanatory comments.

Glutexo commented on app/models.py:
```diff
@@ -326,7 +337,7 @@ class HostSchema(Schema):
     external_id = fields.Str(validate=validate.Length(min=1, max=500))
     facts = fields.List(fields.Nested(FactsSchema))
-    tags = fields.List(fields.Nested(TagsSchema))
+    tags = fields.Raw(allow_none=True)
```
Being benevolent allowed _null_. It is equivalent to an empty array or empty dictionary. It is always [saved](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/app/models.py#L105) as an empty object.

The REST API still [requires](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/swagger/api.spec.yaml#L683) an empty array or a missing _tags_ key. I didn’t change the specification to save work on a soon to be deprecated subsystem.

Glutexo commented on app/models.py:
```diff
@@ -336,6 +347,47 @@ def has_timezone_info(self, timestamp):
+
+    @staticmethod
+    def _validate_tags_list(tags):
+        StructuredTagsSchema(many=True, strict=True).validate(tags)
```
Leveraging the side-effect of _strict=True_. _ValidationError_ is raised if _tags_ doesn’t conform.

Glutexo commented on app/models.py:
```diff
@@ -336,6 +347,47 @@ def has_timezone_info(self, timestamp):
+    def _validate_tags_dict(tags):
+        for namespace, ns_tags in tags.items():
+            TAG_NAMESPACE_VALIDATION(namespace)
+            if ns_tags is None:
+                continue
```
Being benevolent, allowing _null_ instead of the actual sub-object. Why not. The internal representation remains the same – an empty object.

Glutexo commented on app/models.py:
```diff
@@ -336,6 +347,47 @@ def has_timezone_info(self, timestamp):
+
+            for key, values in ns_tags.items():
+                TAG_KEY_VALIDATION(key)
+                if values is None:
+                    continue
```
The same for the list of values. Again, internally always stored as an empty array.

Glutexo commented on app/models.py:
```diff
@@ -336,6 +347,47 @@ def has_timezone_info(self, timestamp):
+                    raise ValidationError("Tag values must be an array or null.")
+
+                for value in values:
+                    if value is None:
+                        continue
```
Allowing none just as empty strings. They’re not going to be stored.

Glutexo commented on app/serialization.py:
```diff
@@ -172,8 +172,67 @@ def _serialize_uuid(u):
+    deserialized = {}
+
+    for tag_data in tags:
+        namespace = Tag.deserialize_namespace(tag_data.get("namespace"))
```
Keeping the logic of treating _"null"_ same as an empty value. I still think that’s not correct (reported as [_RHCLOUD-5803_](https://projects.engineering.redhat.com/browse/RHCLOUD-5803)), but keeping that for consistency.

Glutexo commented on app/serialization.py:
```diff
@@ -172,8 +172,67 @@ def _serialize_uuid(u):
+    deserialized_tags = {}
+
+    for namespace, tags_ns in tags.items():
+        deserialized_namespace = Tag.deserialize_namespace(namespace)
```
Dtto.

Glutexo commented on app/utils.py:
```diff
@@ -352,28 +349,8 @@ def create_tags_from_nested(nested_tags):
         return tags
-
-    @staticmethod
-    def create_structered_tags_from_tag_data_list(tag_data_list):
```
Sneakily stripping flesh from the _Tag_ class, moving it to [_app.serialization_](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/app/serialization.py#L185).

Glutexo commented on swagger/api.spec.yaml:
```diff
@@ -322,13 +322,13 @@ paths:
+                $ref: '#/components/schemas/ActiveTags'
         '400':
-            description: Invalid request.
+          description: Invalid request.
```
Automatic reformatting. 🤷🏻‍♂️

Glutexo commented on test_api.py:
```diff
@@ -239,7 +239,7 @@ def _make_http_call(self, http_method, path, data, status, return_response_as_js
 
     def _response_check(self, response, status, return_response_as_json):
-        self.assertEqual(response.status_code, status)
+        self.assertEqual(status, response.status_code)
```
The first argument should be “expected” and the second “actual”. At least PyCharm’s test suite output thinks so. Swapped those here in these broadly used methods, so if a test fails, the log output of the IDE is closer to the truth.

Glutexo commented on test_api.py:
```diff
@@ -281,7 +281,7 @@ def _build_host_id_list_for_url(self, host_list):
 
     def _verify_host_status(self, response, host_index, expected_status):
-        self.assertEqual(response["data"][host_index]["status"], expected_status)
+        self.assertEqual(expected_status, response["data"][host_index]["status"])
```
Dtto.

Glutexo commented on test_api.py:
```diff
@@ -879,15 +879,79 @@ def test_create_host_with_invalid_ansible_host(self):
+                self._verify_host_status(response, 0, 400)
+
+
+class CreateHostsWithTagsTestCase(DBAPITestCase):
```
Although nothing changed for the REST API, I found the tests not exhaustive. I needed to be sure that nothing broke, so I added some more verifying the actual behavior.

Glutexo commented on test_api.py:
```diff
@@ -879,15 +879,79 @@ def test_create_host_with_invalid_ansible_host(self):
+
+
+class CreateHostsWithTagsTestCase(DBAPITestCase):
+    def test_create_host_with_null_tag_key(self):
```
[_StructuredTag_](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/swagger/api.spec.yaml#L508) doesn’t list _key_ as _required_. That means that it’s ok to omit it, but it’s not to include it, but with _null_ as its value. So this is the only case caught by the request body validation.

rodrigonull replied:
According to the tagging documentation, the "key" should be mandatory.
https://docs.google.com/document/d/1Gfkc3QRWezyHXr8jnrHNYSUsBsetG8bZyx9uCmcRxEg

Glutexo replied:
@rodrigonull Yeah. It’s a bug of our HTTP API specification. Because we are going to remove tags update through HTTP, I didn’t want to change (fix) it. So I just updated the tests so they test the current behavior, making them fail if anything changes unexpectedly.

Glutexo commented on test_api.py:
```diff
@@ -879,15 +879,79 @@ def test_create_host_with_invalid_ansible_host(self):
+        host_data = HostWrapper(test_data(tags=[tag]))
+        self.post(HOST_URL, [host_data.data()], 400)
+
     def test_create_host_with_invalid_tags(self):
```
The length constraints are missing in the API [specification](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/swagger/api.spec.yaml#L508) as well. So these violations are caught by Marshmallow, but not by Connexion.

Glutexo commented on test_db_model.py:
```diff
@@ -194,22 +194,6 @@ def test_host_schema_timezone_enforced():
     assert "Timestamp must contain timezone info" in str(excinfo.value)
 
 
-def test_tag_deserialization():
```
The _create_structered_tags_from_tag_data_list_ is gone, replaced by _ [_app.serialization.:deserialize_tags_list_](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/app/serialization.py#L185). Now exhaustively tested [elsewhere](https://github.com/Glutexo/insights-host-inventory/blob/52fa371dbe0de11cab73eceae73e98eb52aaae15/test_unit.py#L1464).

Glutexo commented on test_mq_service.py:
```diff
@@ -353,8 +356,13 @@ def test_add_host_with_tags(self, datetime_mock):
                 {"namespace": "NS1", "key": "key3", "value": "val3"},
```
Again, trying to make the tests as exhaustive as possible to catch any possible regression.

Glutexo commented on test_api.py:
```diff
@@ -879,15 +879,83 @@ def test_create_host_with_invalid_ansible_host(self):
+
+class CreateHostsWithTagsTestCase(DBAPITestCase):
+    def test_create_host_with_null_tags(self):
+        host_data = HostWrapper(test_data(tags=None))
```
Tags is not listed as _required_, but also not as _nullable_. That means it can be omitted, but when present it can’t be _null_. That’s not correct, but it’s already like that, so added a test to prevent regressions.

jharting commented on app/serialization.py:
```diff
@@ -172,8 +172,67 @@ def _serialize_uuid(u):
+
+        if tag_data["key"] not in deserialized_ns:
+            deserialized_ns[tag_data["key"]] = []
+        deserialized_key = deserialized_ns[tag_data["key"]]
```
Looks like this is more like `deserialized_values` rather than `deserialized_key`

Glutexo replied:
I tried to make the code more readable by shortening the long two-level depp dictionary item references to a single variable. As I see now, it only caused confusion.

Instead, I rather replaced `tag_data["key"]` with just `key`. Like that, even the deepest reference is `deserialized[namespace][key]`, which is not that awful, yet it’s clear it’s a value inside a structure.

jharting commented on utils/payloads.py:
```diff
@@ -513,12 +513,10 @@ def build_host_chunk():
-            {"namespace": "Sat", "key": "prod", "value": None},
-        ],
+        "tags": None,
+        # "tags": {
```
intention or accident?

Glutexo replied:
Accident. 😞

Glutexo replied:
Accident. 😞

Glutexo replied:
The _None_ was an accident. The commented out dictionary version of tags is intentional.

Glutexo commented:
I’ll address the issues mentioned in comments.

Glutexo commented:
I addressed the review comments. Refactored the tags update, so it’s now less fishy and more semantic.

Also I found out that there are no tests for correctly handling legacy hosts with `NULL` or `'null'` in _tags_. Added those.

Glutexo commented:
Found a bug. It’s not possible to delete a _null_ namespace. Working on a test and a fix.

Glutexo commented:
Fixed, tested. Added a few more tests verifying that all no namespace representations (missing key, _null_, empty string and string _"null"_, are treated the same on update. 🎉

Glutexo commented:
🥳
